src/lib.rs

```diff
@@ -156,12 +165,53 @@ impl<T> Deref for Gc<T> {
     }
 }
 
+pub fn init(flags: gc::DebugFlags) {
+    unsafe { COLLECTOR = Some(Collector::new(flags)) };
+}
+
+pub fn collect() {
+    // FIXME: Is this UB? Theoretically this pub `collect` fn can be called in
+    // parallel from multiple threads, allowing us potentially multiple mutable
+    // references to the underlying `COLLECTOR` struct. However, the inner
+    // `collect` method guards against multiple collection invocations via a
+    // mutex on the `CollectorState` field.
+    unsafe { COLLECTOR.as_mut().unwrap().collect() }
+}
+
+pub struct Debug;
+
+impl Debug {
+    pub fn is_black<T>(gc: Gc<T>) -> bool {
+        let collector = unsafe { COLLECTOR.as_ref().unwrap() };
+        let cstate = *collector.state.lock().unwrap();
+
+        // Checking an object colour only makes sense immediately after marking has
+        // taken place. After a full collection has happened, marking results are
+        // stale and the object graph must be re-marked in order for this query
+        // to be meaningful.
+        assert_eq!(cstate, gc::CollectorState::FinishedMarking);
+        return collector.colour(unsafe { Gc::from_raw(gc.objptr as *const i8) })
+            == gc::Colour::Black;
+    }
+
+    /// This should be used with caution. For example, if writing a test using
```

Comment:
This doc string feels a bit confusing. Why not just say something like "If true, this object was not marked for collection in the last sweep"? [BTW, in a sense, your whole warning also applies to `is_black`, since the two functions are simply inverses.]

Reply:
I've reworded this.

I think the warning isn't as relevant to `is_black` as the conservative GC will tend to over-approximate - if I assert an object should be black after a collection and it's not, that's a bug! 

On the other hand, I can't be sure that an object didn't get marked, as for some non-deterministic reason, residual pointers may have been left lingering on the stack or within registers.

Reply:
I think none of this is a problem if you avoid talking about e.g "liveness" (so, for example, it would definitely have been incorrect to say "this guarantees the object was live at the last collection). That said, what can anyone do with such a guarantee? I think the answer is "nothing": by definition, the instant the collection is finished, this information is stale and can't be relied on. So I think you might be worrying unduly :)

Reply:
Sure. I suppose the main point that I am trying to make it clear is that it's safe to use `is_black` in a test, but not `is_white`. It's always safe to over-approximate, and it doesn't really matter if a white object was coloured black as a false positive. However, we can't run a test collection and assert that objects which we didn't expect to be coloured black remain white.

Reply:
> Unless, that is, one explicitly says "you can not safely use !o.is_black()"?

Exactly `!is_black()` is not something you should ever test for. Perhaps I should make that clearer in the doc comments. It's only right to test that a collector reached reachable objects in the last collection.

Reply:
OK, if that's the constraint then stating it explicitly is a good idea. The explanation doesn't have to be long, so long as it's clear.

Reply:
I've decided to get rid of `is_white` altogether to avoid confusion, and warn on using the inverse of `is_black`.